packages/lib-classifier/src/store/TranscriptionReductions/TranscriptionReductions.js

```diff
@@ -0,0 +1,125 @@
+
+const CONSENSUS_SCORE_TO_RETIRE = 3
+const MINIMUM_VIEW_TO_RETIRE = 5
+const REDUCER_KEY = 'alice'
+
+const TranscriptionLine = types.model({
+  consensusReached: types.boolean,
+  frame: types.number,
+  hasCollaborated: types.boolean,
+  points: types.array(types.frozen({})),
+  previousAnnotation: types.boolean,
+  textOptions: types.array(types.string),
+})
+
+const TranscriptionReductions = types
+  .model('TranscriptionReductions', {
+    caesarReducerKey: types.frozen(REDUCER_KEY),
+    error: types.maybeNull(types.frozen({})),
+    frame: types.optional(types.number, 0),
+    loadingState: types.optional(types.enumeration('state', asyncStates.values), asyncStates.initialized),
+    reductions: types.array(types.frozen({})),
+    subjectId: types.string,
+    workflowId: types.string
+  })
+
+  .views(self => {
+    function constructCoordinates (line) {
+      if (line && line.clusters_x && line.clusters_y) {
+        return line.clusters_x.map((point, i) => ({ x: line.clusters_x[i], y: line.clusters_y[i] }))
+      }
+      return []
+    }
+
+    function constructText (line) {
+      const sentences = []
+      if (line && line.clusters_text) {
+        line.clusters_text.forEach(value => {
+          value.forEach((word, i) => {
+            if (!sentences[i]) {
+              sentences[i] = []
+            }
+            if (word && word.length) {
+              sentences[i].push(word)
+            }
+          })
+        })
+      }
+      return sentences.map(value => value.join(' '));
+    }
+
+    function constructLine (annotation, options) {
+      const { minimumViews, threshold } = options
+      const { frame } = self
+      const points = constructCoordinates(annotation)
+      const textOptions = constructText(annotation)
+      return TranscriptionLine.create({
+        points,
+        frame,
+        textOptions,
+        consensusReached:
+          annotation.consensus_score >= threshold ||
+          annotation.number_views >= minimumViews,
+        previousAnnotation: true,
+        hasCollaborated: false,
+      })
+    }
+
+    return {
+      get transcribedLines () {
+        const { frame, reductions } = self
+        let transcribedLines = []
+        reductions.forEach(reduction => {
+          const { parameters } = reduction.data
+          const threshold = parameters?.low_consensus_threshold || CONSENSUS_SCORE_TO_RETIRE
+          const minimumViews = parameters?.minimum_views || MINIMUM_VIEW_TO_RETIRE
+          const currentFrameAnnotations = reduction.data[`frame${frame}`] || []
+          const currentFrameLines = currentFrameAnnotations.map(annotation => {
+            return constructLine(annotation, { minimumViews, threshold })
+          })
+          transcribedLines = transcribedLines.concat(currentFrameLines)
+        })
+        return transcribedLines
+      }
+    }
+  })
+
+  .actions(self => {
+    return {
+      afterAttach () {
+        self.fetchCaesarReductions()
+      },
```

Comment:
I spent the afternoon switching from `afterCreate` to `afterAttach` so that the Caesar client is passed in from the environment. If we switch back, the client will be undefined.

Reply:
This doesn't make sense as the environment should be able to be injected with any model. What were the errors?

Reply:
`undefined.request is not a function`or something like that.

You could change the code and try it yourself. Probably quicker than disturbing me in the middle of the night.

Reply:
By the way, I’ve been testing with workflow 3085 on staging. The reducer key is `ext-17`.

Reply:
When I was working on this earlier, the answer was in the API docs for `getEnv`. https://mobx-state-tree.js.org/API/#getenv

Reply:
![TypeError: Cannot read property 'request' of undefined](https://user-images.githubusercontent.com/59547/75227226-b1a7b580-57a5-11ea-9d82-b81495ed7983.png)
You can't read the tree environment from a model that isn't attached to the tree yet.

Reply:
I had referenced https://mobx-state-tree.js.org/concepts/dependency-injection which was missing the specific bit about requiring attachment. Thanks.

Reply:
I spent a couple of hours on Monday trying to get `afterCreate` to work before the penny dropped about the environment not being available. Every time I think I’ve got my head around MST, it seems to trip me up with something like this.